### PR TITLE
[Tune] Added XGBoost tutorial and template

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a comprehensive XGBoost tutorial to the Tune documentation.

## Related issue number

N/A

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: doc/source/tune/_tutorials/overview.rst [hunk 1]
```diff
@@ -161,6 +162,7 @@ PyTorch Examples
 XGBoost Example
 ~~~~~~~~~~~~~~~
 
+- :ref:`XGBoost tutorial <tune-xgboost>`: A guide to tuning XGBoost parameters with Tune.
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/overview.rst:165]: Can we make this a custom gallery item?

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+Tune supports *fractional GPUs*. This means that each task is assigned a fraction
+of the GPU memory for training. For 10 tasks, this could look like this:
+
+.. code-block:: python
+   :emphasize-lines: 3
+
+    tune.run(
+        train_breast_cancer,
+        resources_per_trial={"cpu": 1, "gpu": 0.1},
+        config=config,
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:497]: Do you have to do anything with the XGBoost model to actually use the GPU?

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+
+The best configuration we found used ``eta=0.0940201``, ``max_depth=5``,
+``min_child_weight=2``, ``subsample=0.711124`` and reached an accuracy of
+``0.972028``.
+
+Efficient trial scheduling
+--------------------------
+Currently, Tune samples 10 different hyperparameter configurations and trains a full
+XGBoost on all of them. In our small example, training is very fast. However,
+if training takes longer, a significant amount of computer resources is spent on trials
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:363]: Maybe rename this to "early stopping". 

This section is also a bit verbose; maybe we can just narrow it down to the key idea (The basic idea of this scheduler: We sample a number of hyperparameter configurations. Each of these configurations is trained for a specific number of iterations. After these iterations, only the best performing hyperparameters are retained. These
are selected according to some loss metric, usually an evaluation loss.) and a short description of the parameters.

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+.. note::
+
+    To run this tutorial, you will need to install the following:
+
+    .. code-block:: bash
+
+        $ pip install xgboost
+
+What is XGBoost
+---------------
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:26]: Is it possible to shorten this section? I think you can cut the region on boosting/bagging/other libraries, and just link readers to another blog post or the XGBoost website.

File: doc/source/tune/_tutorials/tune-xgboost.rst (hunk 2 above)
Comment (krfricke) [reply doc/source/tune/_tutorials/tune-xgboost.rst:497]: Right, you have to set `tree_method="gpu_hist"`. Will update it

File: doc/source/tune/_tutorials/tune-xgboost.rst (hunk 4 above)
Comment (krfricke) [reply doc/source/tune/_tutorials/tune-xgboost.rst:26]: I made it quite a bit shorter and referred to another overview article

File: doc/source/tune/_tutorials/tune-xgboost.rst (hunk 3 above)
Comment (krfricke) [reply doc/source/tune/_tutorials/tune-xgboost.rst]: Done

File: doc/source/tune/_tutorials/overview.rst (hunk 1 above)
Comment (krfricke) [reply doc/source/tune/_tutorials/overview.rst:170]: I added one in the tutorials section

File: doc/source/tune/_tutorials/overview.rst [hunk 5]
```diff
@@ -25,6 +25,11 @@ Take a look at any of the below tutorials to get started with Tune.
    :figure: /images/tune.png
    :description: :doc:`A walkthrough to setup your first Tune experiment <tune-tutorial>`
 
+.. customgalleryitem::
+   :tooltip: Tuning XGBoost parameters.
+   :figure: /images/xgboost_tree.png
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/overview.rst:30]: can we change this to the xgboost logo?

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 6]
```diff
@@ -0,0 +1,504 @@
+.. _tune-xgboost:
+
+Tuning XGBoost parameters
+=========================
+
+XGBoost is currently one of the most popular machine learning algorithms. It performs
+very well on a large selection of tasks, and was the key to success in many Kaggle
+competitions.
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:8]: nit: can you add a small xgboost logo here?

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+        scheduler=scheduler)
+
+Each task thus works with 10% of the available GPU memory. You also have to tell
+XGBoost to use the ``gpu_hist`` tree method, so it knows it should use the GPU.
+
+Conclusion
+----------
+You should now have a basic understanding on how to train XGBoost models and on how
+to tune the hyperparameters to yield the best results. In our simple example,
+Tuning the parameters didn't make a huge difference for the accuracy.
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:501]:
```suggestion
to tune the hyperparameters to yield the best results. In our simple example, tuning the parameters didn't make a huge difference for the accuracy.
```

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+
+Maybe you have noticed the ``config`` parameter we pass to the XGBoost algorithm. This
+is a ``dict`` in which you can specify parameters for the XGBoost algorithm. In this
+simple example, the only parameter we passed is the ``objective`` parameter. The value
+``binary:logistic`` tells XGBoost that we aim to train a logistic regression model for
+a binary classification task. You can find an overview over all valid objectives
+`here in the XGBoost documentation <https://xgboost.readthedocs.io/en/latest/parameter.html#learning-task-parameters>`_.
+
+Configuring XGBoost
+-------------------
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:115]:
```suggestion
XGBoost Hyperparameters
-----------------------
```

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+XGBoosts default parameters already lead to a good accuracy, and even our guesses in the
+last section should result in accuracies well above 90%. However, our guesses were
+just that: guesses. Often we do not know what combination of parameters would actually
+lead to the best results on a machine learning task.
+
+Unfortunately, there are infinitely many combinations of hyperparameters we could try
+out. Should we combine ``max_depth=3`` with ``subsample=0.8`` or with ``subsample=0.9``?
+What about the other parameters?
+
+This is where hyperparameter tuning comes into play. Using tuning libraries such as
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:246]:
```suggestion
This is where hyperparameter tuning comes into play. By using tuning libraries such as
```

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+XGBoost to use the ``gpu_hist`` tree method, so it knows it should use the GPU.
+
+Conclusion
+----------
+You should now have a basic understanding on how to train XGBoost models and on how
+to tune the hyperparameters to yield the best results. In our simple example,
+Tuning the parameters didn't make a huge difference for the accuracy.
+But in larger applications, intelligent hyperparameter tuning can make the
+difference between a model that doesn't seem to learn at all, and a model
+that outperforms all the other ones.
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:504]:
```suggestion
that outperforms all the other ones.

Further References
------------------

(Need to format the below into links)
XGBoost Hyperparameter Tuning - A Visual Guide https://kevinvecmanis.io/machine%20learning/hyperparameter%20tuning/dataviz/python/2019/05/11/XGBoost-Tuning-Visual-Guide.html

Notes on XGBoost Parameter Tuning https://xgboost.readthedocs.io/en/latest/tutorials/param_tuning.html

Doing XGBoost Hyperparameter Tuning the smart way
https://towardsdatascience.com/doing-xgboost-hyper-parameter-tuning-the-smart-way-part-1-of-2-f6d255a45dde
```

**Review by richardliaw — APPROVED** [review]
This looks good to me! Ping me when the comments are addressed and I'll be happy to merge.

Thanks!

File: doc/source/tune/_tutorials/tune-xgboost.rst [hunk 11]
```diff
@@ -501,4 +508,15 @@ to tune the hyperparameters to yield the best results. In our simple example,
...
 difference between a model that doesn't seem to learn at all, and a model
-that outperforms all the other ones.
\ No newline at end of file
+that outperforms all the other ones.
+
+Further References
+------------------
+
+(Need to format the below into links)
+XGBoost Hyperparameter Tuning - A Visual Guide https://kevinvecmanis.io/machine%20learning/hyperparameter%20tuning/dataviz/python/2019/05/11/XGBoost-Tuning-Visual-Guide.html
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-xgboost.rst:517]: Hey @krfricke can you fix this up? 

I think they should be in a format like `Title [link]`.

thanks!
